File: backend/routers/mlops.py
"""
backend/routers/mlops.py
MLOps Center — real MLflow runs + real model file metadata.
"""
import os, sys, json
from datetime import datetime
from fastapi import APIRouter
sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))

router = APIRouter()

MLRUNS_PATH  = os.path.join(os.path.dirname(__file__), "..", "..", "mlruns")
MODEL_PT     = os.path.join(os.path.dirname(__file__), "..", "..", "models", "lstm_best.pt")
METRICS_JSON = os.path.join(os.path.dirname(__file__), "..", "..", "models", "lstm_metrics.json")


def _read_metrics():
    if os.path.exists(METRICS_JSON):
        with open(METRICS_JSON) as f:
            return json.load(f)
    return {}


@router.get("/experiment-runs")
def experiment_runs():
    """
    Read real MLflow file-based tracking from mlruns/ folder.
    If mlruns exists, parse real runs. Otherwise show honest empty state.
    """
    if not os.path.exists(MLRUNS_PATH):
        return {
            "runs": [],
            "note": "No MLflow runs found. Run 'python ml/train_lstm.py' to generate experiment history.",
            "mlruns_exists": False
        }

    try:
        import mlflow
        client = mlflow.tracking.MlflowClient(tracking_uri=f"file://{os.path.abspath(MLRUNS_PATH)}")
        experiments = client.search_experiments()
        all_runs = []
        for exp in experiments:
            runs = client.search_runs(
                experiment_ids=[exp.experiment_id],
                order_by=["start_time DESC"],
                max_results=20
            )
            for r in runs:
                all_runs.append({
                    "run_id":     r.info.run_id[:8],
                    "experiment": exp.name,
                    "status":     r.info.status,
                    "start_time": datetime.fromtimestamp(r.info.start_time / 1000).isoformat() if r.info.start_time else None,
                    "duration_s": round((r.info.end_time - r.info.start_time) / 1000) if r.info.end_time else None,
                    "metrics":    {k: round(v, 4) for k, v in r.data.metrics.items()},
                    "params":     r.data.params,
                    "tags":       dict(list(r.data.tags.items())[:3])
                })
        return {"runs": all_runs, "mlruns_exists": True, "total_runs": len(all_runs)}
    except Exception as e:
        return {
            "runs": [],
            "note": f"MLflow read error: {str(e)}",
            "mlruns_exists": os.path.exists(MLRUNS_PATH)
        }


@router.get("/model-registry")
def model_registry():
    """Return real model file info — honest about what's actually trained."""
    models = []

    if os.path.exists(MODEL_PT):
        stat = os.stat(MODEL_PT)
        metrics = _read_metrics()
        models.append({
            "name":          "LSTM Demand Forecaster",
            "version":       "1.0",
            "stage":         "Production",
            "file":          "models/lstm_best.pt",
            "file_size_mb":  round(stat.st_size / 1024 / 1024, 3),
            "last_modified": datetime.fromtimestamp(stat.st_mtime).isoformat(),
            "metrics":       metrics,
            "architecture":  "LSTM (2 layers, hidden=64, seq_len=14)",
            "note":          "Model retraining in progress — MAPE improvement expected."
        })

    metrics_exists = os.path.exists(METRICS_JSON)
    return {
        "models": models,
        "total": len(models),
        "note": "Only models that have been physically trained and saved are shown. No fabricated history.",
        "metrics_json_exists": metrics_exists,
        "model_pt_exists": os.path.exists(MODEL_PT)
    }


@router.get("/dag-status")
def dag_status():
    """Return real pipeline step status based on actual file existence."""
    steps = [
        {"name": "Data Ingestion",       "file": "datasets/raw",                           "status": None},
        {"name": "Data Processing",      "file": "datasets/processed",                     "status": None},
        {"name": "Feature Engineering",  "file": "ml/data_loader.py",                      "status": None},
        {"name": "Model Training",       "file": "models/lstm_best.pt",                    "status": None},
        {"name": "Metrics Logging",      "file": "models/lstm_metrics.json",               "status": None},
        {"name": "Market Scheduler",     "file": "datasets/processed/live_prices.json",    "status": None},
        {"name": "Alert System",         "file": "scripts/email_alerter.py",               "status": None},
        {"name": "API Backend",          "file": "backend/main.py",                        "status": None},
    ]
    base = os.path.join(os.path.dirname(__file__), "..", "..")
    for s in steps:
        path = os.path.join(base, s["file"])
        s["exists"] = os.path.exists(path)
        s["status"] = "success" if s["exists"] else "pending"
        if s["exists"] and os.path.isfile(path):
            stat = os.stat(path)
            s["last_modified"] = datetime.fromtimestamp(stat.st_mtime).isoformat()
        del s["file"]
    return steps

File: frontend/src/lib/mock-data/mlops.ts
export const experiments = [
  { id: 'EXP-015', name: 'TFT Multi-Region v4', model: 'TFT', status: 'running', mape: 0, mae: 0, rmse: 0, params: { epochs: 200, lr: 0.001, batchSize: 64 }, startTime: '2h ago', duration: '-', tags: ['production', 'v4'] },
  { id: 'EXP-014', name: 'LSTM Demand Forecaster', model: 'LSTM', status: 'completed', mape: 5.82, mae: 67.4, rmse: 89.2, params: { epochs: 150, lr: 0.002, batchSize: 32 }, startTime: '1d ago', duration: '4.2h', tags: ['staging'] },
  { id: 'EXP-013', name: 'GRU Seasonal', model: 'GRU', status: 'completed', mape: 6.14, mae: 72.1, rmse: 94.8, params: { epochs: 120, lr: 0.001, batchSize: 64 }, startTime: '2d ago', duration: '3.1h', tags: ['experiment'] },
  { id: 'EXP-012', name: 'Ensemble Stacking', model: 'XGBoost', status: 'completed', mape: 5.94, mae: 69.8, rmse: 91.4, params: { epochs: 300, lr: 0.01, batchSize: 128 }, startTime: '3d ago', duration: '2.8h', tags: ['production'] },
  { id: 'EXP-011', name: 'XGBoost Baseline', model: 'XGBoost', status: 'failed', mape: 0, mae: 0, rmse: 0, params: { epochs: 500, lr: 0.01, batchSize: 256 }, startTime: '4d ago', duration: '0.5h', tags: ['failed'] }
];

export const modelRegistry = [
  { id: 'MOD-005', name: 'LSTM Demand Forecaster', version: 'v2.4.1', stage: 'Production', mape: 5.82, createdAt: 'Jun 15, 2025', updatedAt: 'Jun 16, 2025' },
  { id: 'MOD-004', name: 'TFT Multi-Region', version: 'v2.4.0', stage: 'Staging', mape: 5.94, createdAt: 'Jun 12, 2025', updatedAt: 'Jun 13, 2025' },
  { id: 'MOD-003', name: 'GRU Seasonal', version: 'v2.3.8', stage: 'Staging', mape: 6.14, createdAt: 'Jun 8, 2025', updatedAt: 'Jun 9, 2025' },
  { id: 'MOD-002', name: 'Ensemble Stacking', version: 'v2.3.5', stage: 'Archived', mape: 6.48, createdAt: 'Jun 1, 2025', updatedAt: 'Jun 2, 2025' },
  { id: 'MOD-001', name: 'LSTM Baseline', version: 'v2.2.0', stage: 'Archived', mape: 8.34, createdAt: 'May 20, 2025', updatedAt: 'May 21, 2025' }
];

export const dagSteps = [
  { id: 'DAG-1', name: 'Data Ingestion', status: 'success', duration: '45m', lastRun: '2h ago' },
  { id: 'DAG-2', name: 'Feature Engineering', status: 'success', duration: '28m', lastRun: '2h ago' },
  { id: 'DAG-3', name: 'Model Training', status: 'running', duration: '-', lastRun: 'started 2h ago' },
  { id: 'DAG-4', name: 'Evaluation', status: 'pending', duration: '-', lastRun: '-' },
  { id: 'DAG-5', name: 'Staging Deploy', status: 'pending', duration: '-', lastRun: '-' },
  { id: 'DAG-6', name: 'A/B Test', status: 'pending', duration: '-', lastRun: '-' },
  { id: 'DAG-7', name: 'Production Deploy', status: 'pending', duration: '-', lastRun: '-' }
];

export const monitoringData = {
  mapeOverTime: Array.from({length: 30}).map((_, i) => ({ date: `Jun ${i+1}`, value: 6.5 - (i * 0.02) + (Math.random() * 0.3) })),
  driftScore: Array.from({length: 30}).map((_, i) => ({ date: `Jun ${i+1}`, value: 0.1 + (Math.random() * 0.15) + (i === 15 ? 0.2 : 0) }))
};
